### ISSUE TITLE
runtime: Go program "just terminates" with exit 0

### ISSUE THREAD

**Issue by purpleidea** [golang/go#15873]
While hacking on a medium sized golang project, I've found that certain things seem to cause my program to terminate, but without any trace or exit code. I'm assuming this is a golang bug, since there's no clear reason why the program would exit without reason.

I've got a binary that I compiled that experiences this, and I can reproduce it on other machines.

I'm running golang:
go version go1.5.4 linux/amd64
On Fedora 23.

If this is a known issue someone can point me to, I'd appreciate it, otherwise please let me know what I can run to debug this.

Here is the output of my most recent terminate:

```
./mgmt run
06:12:56 main.go:65: This is: mgmt, version: 0.0.3-48-ga89c332-dirty
06:12:56 main.go:66: Main: Start: 1464516776777766845
06:12:56 main.go:120: Main: Seeds: No seeds specified!
06:12:56 etcd.go:344: Etcd: Bootstrapping...
06:12:56 etcd.go:1089: Trace: Etcd: nominateCallback()
06:12:56 asm_amd64.s:438: Trace: Etcd: nominateCallback(): Finished!
$ echo $?
0
```

Of particular note is that the last line contains the contents of a log.Printf command, which is some debugging text I put there, but it doesn't happen at asm_amd64.s:438 (I've enabled logging of file and line numbers)

Thanks for your time, and I'm happy to post the binary if anyone is interested.

**Comment by minux** [comment]
The runtime shouldn't exit with status 0,
there must be some code that is caling
os.Exit(0).

Please double check the source code,
and if you're using cgo, check the external
c/c++ code too.

**Comment by purpleidea** [comment]
@minux Of course I agree with you, otherwise I wouldn't have reported this. There's nothing I'm calling that calls Exit, and the etcd authors said there is nothing that does it either. When it happened an earlier time I figured it was my fault, but when I got the log with the assembly code, I realized it was probably a golang bug. There's no reason I would see that asm line AFAIK.

**Comment by davecheney** [comment]
It's pretty hard to diagnose this without the source of your program. Can
you please provide a minimal example that demonstrates the problem.

**Comment by purpleidea** [comment]
On Sun, May 29, 2016 at 6:11 PM, Dave Cheney <notifications@github.com>
wrote:

> It's pretty hard to diagnose this without the source of your program. Can
> you please provide a minimal example that demonstrates the problem.
>


Once I can distill the code down, I will. I suppose the binary won't help?
Does the ASM message tell you anything?

**Comment by minux** [comment]
You could try using gdb to debug the program and
use "catch syscall exit_group" to catch the point
where the process issues the exit syscall.

When the catchpoint is hit, please issue bt command
to see exactly where issues the exit syscall.

**Comment by purpleidea** [comment]
This seems to be related to issues where a null pointer would get used and in the protobuf golang package. So in theory if your code has zero bugs (mine doesn't) then you won't hit this. So something is either failing in a strange way, or something relating to protobufs is suppressing the trace dump on null pointer.

**Comment by ianlancetaylor** [comment]
We need a test case or the backtrace that minux suggested.

**Comment by awishformore** [comment]
I am running into the exact same problem in one of our microservices. The only thing we were able to observe is that some network polling fails before it happens, but that might or might not be related.

I searched the entire code base (including vendored libraries), and the only place where it calls `os.Exit(0)` is an the end of the main function. Before it exits, there are a few lines of log output which are never written.

It simply shouldn't be possible for the application to exit with `code 0` and have no output at all, other than normal operation.

**Comment by ianlancetaylor** [comment]
We need a test case.

**Comment by awishformore** [comment]
Our issue was, in fact, unrelated to this. The shutdown messages were just hidden in front of hundreds of warnings in the log, as execution was terminating and closing everything down. The reason for the shutdown was an unexpected signal other than SIGINT, SIGTERM and SIGHUP being issues, and us deciding to catch all signals. After we stopped catching the signal, the software stopped "crashing". We still don't know what issued the signal, since we found nothing in OS logs.
